pkg/operator/controller/controller.go

```diff
@@ -125,6 +151,24 @@ func (r *ReconcileCDI) Reconcile(request reconcile.Request) (reconcile.Result, e
 		return reconcile.Result{}, err
 	}
 
+	//compare target and observed
+	//Retriveed CR will contain previous version ImageTag and ImageRegistry
+	//while namespaced arguments will contain new version
+	//if versions are not the same: update TargetVersion in CDI cr and move to implementation
+	isUpgrade, err := shouldTakeUpdatePath(r.namespacedArgs.DockerTag, cr.Status.ObservedVersion)
```

Comment:
maybe check if we are upgrading after we check `cr.DeletionTimestamp`?

Reply:
ok, I switched between Deletion check and upgrade, but this is racy

Reply:
I think we have to handle delete during upgrade.  The possible issue I see is that we have to make sure to delete orphaned resources from N-1 install.

Reply:
@mhenriks added logic that call lcleanupUnusedResources if DeletionTimestamp is detectd during upgrade

Reply:
What type of race condition are you concerned about?